### ISSUE TITLE
Clarifications in training config

### ISSUE THREAD

**Issue by Remorax** [PANXiao1994/mRASP2#19]
Hello

Thank you for your excellent work, and well-documented repo. I am trying to use your code to train a new model from scratch, and require some clarification on certain parts that are unclear to me, especially regarding the config. 

(Please note I am referring to this [config](https://github.com/PANXiao1994/mRASP2/blob/new_impl/examples/configs/parallel-12e12d-emb1024-contrast-ras.yml) on the `new_impl` branch as an example of how I could create my own)

1. `data` (under `meta`):  What does this refer to? Is this the directory that contains binarized versions of one multilingual parallel dataset made by concatenating datasets from several language pairs (eg. en-es, en-fr, en-it) or does it contain language pair-specific binary files in its subdirectories? 

2. I can see that in [load_config.sh](https://github.com/PANXiao1994/mRASP2/blob/new_impl/scripts/load_config.sh) variables starting with `meta_` are not written to the `options` variable, and both monolingual and parallel data are provided separately in `train_multilingual_w_mono.sh`. This seems to suggest that paths are expected in the form of `data_1`, `data_2` etc.. If so, could you please confirm what these paths refer to? I.e. how does `data_1` differ from `data_2`?

3. What is `mono_dae`? This is referred to repeatedly in the codebase, at various places. Would I need to set `mono_key` in the config file to `mono_dae`?

4. Lastly, I have parallel and monolingual datasets that I have already preprocessed (with RAS substitution and language token prefixes). Would I need to set variables like `langtoks`, `encoder_langtok` and `decoder_langtok`?

Hope I can receive some assistance on this issue soon. Thanks!

**Comment by PANXiao1994** [comment]
Hi, Sorry for replying late! 
1. `meta_data` refers the parallel training data, it contains language pair-specific binary files, but not in its subdirectories, instead, they should have filenames like "train.en-fr.en.idx" and "train.en-fr.en.bin"
2. `data_1` and `data_2` are actually concatenated during training. This is to avoid having to concatenate dataset manually: you can have multiple training set binarized and you don't need to concatenate them and re-binarize.
3. `mono_dae` refers to the auto-encoded data, where the source sentence will be randomly masked (like BERT).  Since we don't need auto-encoded noise, you should set `mono_key: mono` or any key you like, as long as you keep `langtoks:`(`{\"mono\":(\"src\",\"tgt\")}`) in the config file and `mono_key` consistent.
4. new_impl version implements on-line RAS, so that means you don't need to do RAS operation as a preprocessing step. You should set `langtoks`(for monolingual data), and `encoder_langtok`, `decoder_langtok`(for parallel data)

**Comment by Remorax** [comment]
Hello, no worries, thanks for your detailed answer!

I now understand the codebase a lot better. I have been using the `new_impl` branch for training my model due to a few factors. I wish to conduct all preprocessing offline though (including prepending the language tokens and RAS). What I have done, therefore, as a hack is to pass a flag on [line 647](https://github.com/PANXiao1994/mRASP2/blob/new_impl/mcolt/data/multilingual/multilingual_data_manager.py#L647) of `multilingual_data_manager.py`, to indicate whether or not RAS should be conducted. If RAS is not to be conducted, it will treat it as a dataset with prepended langtoks (since I have already done so in the preprocessing stage). Do you think this should work, given your experience in dealing with fairseq code? Model training seems to work fine with this script.

The other thing I wanted to clarify is whether I can use `translation_w_langtok` while evaluating a model trained using the `translation_multi_simple_epoch_mcolt` task? As far as I can see, the only major difference appears to be that the former deals with binarized test sets while the latter batches raw text on the fly - similar to fairseq-generate vs fairseq-interactive. Am I correct in this assumption?

**Comment by PANXiao1994** [comment]
Hi Remorax,

The first question is too detailed, have you tried and succeeded? If you want to conduct RAS offline, you can remove RAS-related configs(ras_dict, ras_replace_prob) in the config file.

For the second question, yes, you can use n use translation_w_langtok while evaluating a model trained using the translation_multi_simple_epoch_mcolt task. They are not fairseq-generate vs fairseq-interactive, but two tasks with different arguments. They actually do the same thing: prepend a language token to a sentence (both source and target side).

**Comment by Remorax** [comment]
Thank you for answering my questions, Xiao. I will be closing this issue now.

Congrats on your excellent work!
